### PR TITLE
feat - add draft ru am adaptation recipe

### PR DESCRIPTION
**Pull request by SanzharMrz**

### CODE REVIEW

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+*wav.scp*
+                                       
+![wav.scp](https://user-images.githubusercontent.com/48170101/117793265-e0586180-b26d-11eb-96d3-4614ed6c66c7.png)
+
+*utt2spk*
+
+![utt2spk](https://user-images.githubusercontent.com/48170101/117793486-17c70e00-b26e-11eb-8104-9f13f35ca259.png)
+
+*spk2gender*
```
Comment (nshmyrev) [inline vosk-am-adaptation-ru/am-adaptation-ru-model.md]: spk2gender не нужен

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+# Create features
+
+The steps/scripts from the [CVTE finetuning](https://github.com/zhaoyi2/CVTE_chain_model_finetune/tree/master/steps) recipe are used to extract mel features and normalize them, **MAYBE** we should use the steps from kaldi [aishell2](https://github.com/kaldi-asr/kaldi/tree/master/egs/aishell2/s5), can this __affect__ the preparation of features? As for the finetune script, they are combined into one sh [finetune_tdnn_1a.sh](https://github.com/kaldi-asr/kaldi/blob/master/egs/aishell2/s5/local/nnet3/tuning/finetune_tdnn_1a.sh), this pipeline is recommended in alphacephei [model adaptation](https://alphacephei.com/vosk/adaptation) 
+
+Here is the script we use, [ft.sh](ft.sh). Directly a piece of code with data preparation (possible inaccuracies, worth validating):
+
+```bash
+# Compute mfcc 
+steps/make_mfcc.sh \
```
Comment (nshmyrev) [inline vosk-am-adaptation-ru/am-adaptation-ru-model.md:109]: два раза извлекать не надо, один раз признаки извлекаются с конфигом vosk-model-ru-0.10/conf/mfcc.conf

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  ${data_dir}_hires exp/make_mfcc/${data_set}_hires mfcc_hires
+
+steps/compute_cmvn_stats.sh ${data_dir}_hires exp/make_mfcc/${data_set}_hires mfcc_hires
+
+# Compute ivector
+sh steps/online/nnet2/extract_ivectors_online.sh $data_dir ivector exp/nnet3_online/ivectors_test
+
+# Align with ivector
+# By the way, in align.sh there is our modification on the size of the beam, in order to maintain as many hypotheses as possible# beam=1000
+# retry_beam=10000
```
Comment (nshmyrev) [inline vosk-am-adaptation-ru/am-adaptation-ru-model.md]: beam не надо менять, оставить как есть.

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+sh steps/nnet3/align.sh $data_dir data/lang $ali_dir exp/nnet3/ali
```
Comment (nshmyrev) [inline vosk-am-adaptation-ru/am-adaptation-ru-model.md]: ivector отсутствует

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+# Moving the counted ali
+cp exp/nnet3/ali/ali.1.gz $ali_dir/ali.1.gz
+```
+
```
Comment (nshmyrev) [inline vosk-am-adaptation-ru/am-adaptation-ru-model.md:129]: Пропущен этап с извлечением альтернативных транскрипций (latgen)

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+make
+
+# Download and unzip model
+cd ../
+wget https://alphacephei.com/vosk/models/vosk-model-ru-0.10.zip
+unzip vosk-model-ru-0.10.zip
+```
+
+# Prepare data
+The official guide from Kaldi [Data Preparation](https://kaldi-asr.org/doc/data_prep.html).  There is also a code from JohnDoe, for [RM](https://catalog.ldc.upenn.edu/LDC93S3C) dataset, where [prepare_data.py](https://github.com/JohnDoe02/kaldi/blob/private/egs/rm/s5/local/prepare_data.py) parses
```
Comment (nshmyrev) [inline vosk-am-adaptation-ru/am-adaptation-ru-model.md]: Данные надо разбить на train/valid чтобы проверять адаптацию

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+With the following parameters:
+```bash
+num_jobs_initial=1
+num_jobs_final=1
+num_epochs=5
+initial_effective_lrate=0.000005
+final_effective_lrate=0.000001
+minibatch_size=128
+```
+![training](https://user-images.githubusercontent.com/48170101/117966891-43b5c280-b346-11eb-8cd4-3420b72852e9.png)
```
Comment (nshmyrev) [inline vosk-am-adaptation-ru/am-adaptation-ru-model.md:161]: После адаптации надо измерять точность

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 2 above)
Comment (dairm) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md:109]: То есть нужно оставить одно извлечение, либо mfcc_hires либо mfcc?  и при обучении использовать что то одно

Comment (nshmyrev) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md:109]: Не либо либо а просто одно. steps/make_mfcc.sh --conf=model/conf/mfcc.conf

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 5 above)
Comment (dairm) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md:129]: решетки графа ali.1.gz формирует скрипт aligh.sh из папки steps/nnet3, а файл align_lat.sh формирует lattices, если я правильно понял данный скрипт нужно тоже добавить в pipeline

Comment (dairm) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md:129]: еще вопрос по frame_subsampling_factor(в русскоязычной модели он равен 3), в скриптах alignment он часто используется, мне кажется что данный параметр выставляется для ускорения процесса, поэтому можно его оставить 1, пока тестируем

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 4 above)
Comment (SanzharMrz) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md]: путь к ivector указан внутри align.sh online_ivector_dir=exp/nnet3_online/ivectors_test
подскажите, насколько корректно использовать скрипт steps/online/nnet2/extract_ivectors_online.sh

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 1 above)
Comment (SanzharMrz) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md]: принято

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 3 above)
Comment (dairm) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md]: пытаемся запустить aligh.sh и align_lats.sh со стандартным beam=10, retry_beam=40, при генерации ali.1.gz и lattices падает в ошибку, 
Логи приложил
[generate_lattices.1.log](https://github.com/alphacep/vosk-space/files/6506212/generate_lattices.1.log)
[align.1.log](https://github.com/alphacep/vosk-space/files/6506263/align.1.log)

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 7 above)
Comment (SanzharMrz) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md:161]: принято, опишу подсчет wer и прикреплю этот скрипт для логирования предсказаний

[inference.txt](https://github.com/alphacep/vosk-space/files/6506318/inference.txt)

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 3 above)
Comment (nshmyrev) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md]: acoustic-scale должно быть 1.0, frame-subsampling-factor 3. Надо создать файл frame_subsampling_factor с 3 внутри в каталоге модели, тогда скрипты автоматически его подцепят.

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 4 above)
Comment (nshmyrev) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md]: > подскажите, насколько корректно использовать скрипт steps/online/nnet2/extract_ivectors_online.sh

корректно

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 3 above)
Comment (dairm) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md]: спасибо, сейчас попробуем с этими параметрами запустить

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 2 above)
Comment (SanzharMrz) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md:109]: Done.

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 7 above)
Comment (SanzharMrz) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md:161]: 1. https://github.com/SanzharMrz/vosk-space/commit/3a76a336ee5b1c01a9401c2c7c4a38db8b2dadc6
2. https://github.com/SanzharMrz/vosk-space/commit/2adeb6c78e590c0eac7a660eefbc0b8c7e46178b

Done.

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 2 above)
Comment (nshmyrev) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md:109]: Не done. conf каталог откуда должен взяться? Сами то свой скрипт запускали?

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md (hunk 7 above)
Comment (nshmyrev) [reply vosk-am-adaptation-ru/am-adaptation-ru-model.md:161]: Где это в скрипте?

File: vosk-am-adaptation-ru/am-adaptation-ru-model.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    --cmd "$train_cmd" --nj $nj --mfcc-config conf/mfcc.conf \
+    ${data_dir} exp/make_mfcc/${data_set} mfcc
+    
+# Normalization
+steps/compute_cmvn_stats.sh ${data_dir} exp/make_mfcc/${data_set} mfcc || exit 1;
+
+utils/fix_data_dir.sh ${data_dir} || exit 1;
+
+# Compute ivector
+sh steps/online/nnet2/extract_ivectors_online.sh $data_dir ivector ivector_dir
```
Comment (nshmyrev) [inline vosk-am-adaptation-ru/am-adaptation-ru-model.md]: $ должен быть? $ivector_dir?

**Comment by nshmyrev** [comment]
I tried your last code, the script fails with the following error:

```
finetuning-am-ru.sh: line 10: ./cmd.sh: No such file or directory
-----
2. Prepare data.
-----
Creating Kaldi format data from ru_RU/by_book/male/minaev/oblomov/ ...
python3: can't open file '../create_data.py': [Errno 2] No such file or directory
sh: 0: Can't open ../dict_prep.sh
finetuning-am-ru.sh: line 79: utils/prepare_lang.sh: No such file or directory
```

I think you have stage 3 hardcoded so it doesn't work out of box. Please fix the script to run from clean state and make sure all dependencies are provided.
